### ISSUE TITLE
Image Data Set Yielding 10% Accuracy

### ISSUE THREAD

**Issue by DennisFaucher** [uber/ludwig#242]
**Describe the bug**
What would the proper YAML be for the MNIST character font image sample data set? I am using:
input_features:
    -
        name: char_path
        type: image
output_features:
    -
        name: char_class
        type: category

and getting 10% accuracy

**To Reproduce**
See Google Colab notebook here: https://colab.research.google.com/drive/16DT1frYPCGKQCMN63dWTWdj2rpcTqLBM 

Please provide code, yaml definition file and a sample of data in order to entirely reproduce the issue.
Issues that are not reproducible will be ignored.

**Expected behavior**
Expected high accuracy

**Screenshots**
![Screen Shot 2019-03-23 at 9 44 23 PM](https://user-images.githubusercontent.com/9034190/54873796-df75c480-4db4-11e9-9554-ba499f6e71d0.png)

**Environment (please complete the following information):**
 - OS: [e.g. iOS] Google Colab Linux
- Python version: 3.6
- Ludwig version: Latest

**Additional context**
Thank you

**Comment by gobzer** [comment]
First, it's **not** a famous MNIST database of handwritten digits.
When people talk about MNIST database, they usually mean this:
http://yann.lecun.com/exdb/mnist/
 - it consists of **handwritten digits**
 - 28 x 28 images
 - 60 000 examples
 - 10 000 more examples (test data)

Your data:
 - digits, printed with **different fonts**
 - 71 x 77
 - ~5 000  examples
 - ~5 000 unlabeled

**Comment by DennisFaucher** [comment]
@gobzer Thank you. I'm new at this. Yes, this data sets is fonts not handwritten characters. I used a simple YAML and trained the font data over 20 hours at 10% accuracy. This example Python [code](https://www.digitalocean.com/community/tutorials/how-to-build-a-neural-network-to-recognize-handwritten-digits-with-tensorflow) trained handwritten digits in a few minutes at 90+% accuracy. I guess more options in the YAML would help with Ludwig, I'm just not sure which options to use.

**Comment by gobzer** [comment]
Here is the script that downloads original MNIST database and converts it to images and CSV:
https://gist.github.com/gobzer/dda71ee2e0be279050f87c6e75fb16c3
I was looking at the tensorflow tutorial https://www.tensorflow.org/tutorials
and made this model:
**m2.yaml**
```
input_features:
    -
        name: path
        type: image
        encoder: stacked_cnn # default
        fc_size: 512 # default 256
        dropout: 0.2 # default false
        filter_size: 0 # default 3
        num_filters: 0 # default 256
        
output_features:
    -
        name: digit
        type: category

```
After 1 hour of training on CPU (1 epoch):
![image](https://user-images.githubusercontent.com/12876046/54994416-7f556d00-4fc4-11e9-8a7f-1d9700e17797.png)

The result is better, but it still not that 98%. And the training runs much slower.

**Comment by DennisFaucher** [comment]
@gobzer Thank you very much for taking the time to explain this to me. I need to study up quite a bit more on the fundamentals of ML and which knobs to turn for success. Ludwig removes the programming aspect of ML but does not remove the need to understand neural networks.

**Comment by gobzer** [comment]
@DennisFaucher the truth is: **I have no idea what I did**

**Comment by DennisFaucher** [comment]
@gobzer I created a Google Colab so that others may benefit from your explanation: https://colab.research.google.com/drive/1XTvuc210e9sfiqzUDWUpn29xWwW-mOtg
Running it now...

**Comment by w4nderlust** [comment]
Wow, thank you @gobzer for helping out! I'm changing the title as it is misleading.
Anyway, we were able to get 98%+ accuracy on mnist with just a little change to the default parameters. Can't recollect the exact ones, but if needed I can find them.

"Ludwig removes the programming aspect of ML but does not remove the need to understand neural networks." yeah that is true. You can easily train a model to decent accuracy in most cases, but to get something that works really well, you need to know what knobs to turn :)

**Comment by gobzer** [comment]
I trained a new model. It performs better:

- orange - base
- blue - m2
- red - m3

![image](https://user-images.githubusercontent.com/12876046/55077918-328f9600-5099-11e9-9c8f-db69d5fd9427.png)

I did this visualisation with:
`
tensorboard --logdir m1:experiment_run_0/model/log,m2:experiment_m2_1/model/log,m3:experiment_m3_0/model/log
`

### m3.yaml

```
# tinkering with MNIST model number 3 
input_features:
    -
        name: path
        type: image
        encoder: stacked_cnn # default
        fc_size: 512 # default 256
        dropout: 0.2 # default false
        filter_size: 0 # default 3
        num_filters: 0 # default 256
        
output_features:
    -
        name: digit
        type: category
        activation: softmax # default relu
        fc_size: 10 # default 256

```

But still. I feel like:
https://www.youtube.com/watch?v=12LLJFSBnS4
I started to read http://neuralnetworksanddeeplearning.com/ and it seems promising. But NN heavily relies on maths, and I'm having difficulties resolving these cognitive dependencies.

> if needed I can find them.

@w4nderlust I need them =)

But there is a thing I'm really concerned about. The tf/keras MNIST tutorial runs in minutes (300 000 steps), and ludwig trains one step in 11 or 12 seconds on the same machine. The performance difference is HUGE!!!

**Comment by DennisFaucher** [comment]
This is great. Thank you @gobzer. I'll try this today.

**Comment by w4nderlust** [comment]
@gobzer there is no substantial difference in training time. Ludwig saves the model twice by default (one if validation improves and one for keeping progress if needed) if you remove these 2 saves speed should be comparable.
Moreover, the default conv stack is probably overparametrized for mnist, so the model you are training is likely bigger than needed.
Finally: `filter_size` and `num_filters` should be values greater than zero, and the final `activation: softmax` and `fc_size``: 10` are not needed. Ludwig automatically adds a layer like that at the end.

**Comment by DennisFaucher** [comment]
Sometimes these runs seem to differ from machine to machine. I just used your YAML in Google Colab and ended up with 11% accuracy.

Colab: https://drive.google.com/open?id=1XTvuc210e9sfiqzUDWUpn29xWwW-mOtg 

YAML:
input_features:
    -
        name: digit_image
        type: image
        encoder: stacked_cnn
        fc_size: 512
        dropout: 0.2
output_features:
    -
        name: digit
        type: category
training:
        epochs: 100

Results:
<img width="500" alt="Screen Shot 2019-03-27 at 9 51 27 PM" src="https://user-images.githubusercontent.com/9034190/55123989-06950480-50db-11e9-86e4-21080eee1a82.png">

**Comment by gobzer** [comment]
@DennisFaucher I checked your notebook and found a mistake.
You use test data for training.
`!cat /content/testing/data.csv | sed '2,$s/^\"/\"\/content\/testing\//' > data2.csv`
In `MNIST dataset` test data is a portion of training data. Ludwig automatically slices data into training/test/validation. So, within my experiments I don't use test data at all.

Here is my python notebook.
https://colab.research.google.com/drive/1lekKocEQRq5Owon5CIoSB3lBY1EOur3C

I think that only **dropout** really matter. If you can improve it, leave a note.

**Comment by DennisFaucher** [comment]
@gobzer That would do it🤦‍♂️
I'll try this again today. Thank you.

**Comment by w4nderlust** [comment]
In Ludwig v0.1.1 we changed the default parameters of the cnn stack encoder, plus we added an [example](https://uber.github.io/ludwig/examples/#image-classification-mnist) on how to train a simple image classifier on MNIST that gets high accuracy. Closing this thread.

**Comment by DennisFaucher** [comment]
@w4nderlust Wow, the examples page has filled out nicely since Ludwig was first released. Nice examples to assist with learning and which knobs to turn for each type of experiment.

**Comment by w4nderlust** [comment]
Thank you!
